cycpp bug

I am getting errors when generating for an agent that I wasn't getting a few days ago.  My agent's code and build system have not changed during this period.  Here is the traceback:

```
cycpp --cpp-path=/usr/bin/clang++ --pass3-use-orig separator.h -o separator.gen.h
Traceback (most recent call last):
  File "/home/r/cyc/install/bin/cycpp.py", line 1663, in <module>
    main()
  File "/home/r/cyc/install/bin/cycpp.py", line 1655, in main
    newfile = generate_code(canon if ns.pass3_use_pp else orig, context, superclasses)
  File "/home/r/cyc/install/bin/cycpp.py", line 1454, in generate_code
    cg.generate(statement, sep)
  File "/home/r/cyc/install/bin/cycpp.py", line 1421, in generate
    transformed = filter.transform(statement, sep)
  File "/home/r/cyc/install/bin/cycpp.py", line 1338, in transform
    rtn += f.transform(statement, sep)
  File "/home/r/cyc/install/bin/cycpp.py", line 731, in transform
    impl = self.impl(ind=ind * " ")
  File "/home/r/cyc/install/bin/cycpp.py", line 1169, in impl
    k_type = self._type(t[1], schematype[0])
TypeError: 'NoneType' object has no attribute '__getitem__'
Makefile:85: recipe for target 'separator.gen.h' failed
make: *** [separator.gen.h] Error 
```

The agent is in the "separator" in https://github.com/rwcarlsen/transoptim/tree/master/agents - the other agents in that folder seem to still be generating fine.  These agents require github.com/cyclus/kitlus to be installed.

@scopatz - any ideas?

--- a/cli/cycpp.py
+++ b/cli/cycpp.py
@@ -1165,7 +1165,8 @@ class SchemaFilter(CodeGeneratorFilter):
                     impl += i + '"{0}<element name=\\"val\\">\\n"\n'.format(xi.up())
                     impl += i + '"{0}<data type=\\"{1}\\" />\\n"\n'.format(xi, el_type)
                     impl += i + '"{0}</element>\\n"\n'.format(xi.down())
-                else: # map
+                else:  # map
+                    schematype = [None, None] if schematype is None else schematype
                     k_type = self._type(t[1], schematype[0])
                     v_type = self._type(t[2], schematype[1])
                     impl += i + '"{0}<element name=\\"key\\">\\n"\n'.format(xi.up())
@@ -1183,6 +1184,7 @@ class SchemaFilter(CodeGeneratorFilter):
                 impl += i + '"{0}<data type=\\"{1}\\" />\\n"\n'.format(xi, d_type)
                 impl += i + '"{0}</element>\\n"\n'.format(xi.down())
             elif t[0] == 'std::pair':
+                schematype = [None, None] if schematype is None else schematype
                 f_type = self._type(t[1], schematype[0])
                 s_type = self._type(t[2], schematype[1])
                 impl += i + '"{0}<element name=\\"{1}\\">\\n"\n'.format(xi.up(), member)

--- a/tests/cycpp_tests.py
+++ b/tests/cycpp_tests.py
@@ -212,7 +212,7 @@ class MockCodeGenMachine(object):
         self.depth = 0
         self.execns = {}
         self.context = {"MyFactory": OrderedDict([('vars', OrderedDict([ 
-            ('x', {'type': 'int'}), 
+            ('x', {'type': 'int', 'schematype': 'positiveInteger'}), 
             ('y', {'type': 'std::string',
                    'shape': [42],
                    'initfromcopy': 'y=m -> y;\n',
@@ -325,9 +325,34 @@ def test_schemafilter():
     exp_impl = ('  return ""\n'
                 '    "<interleave>\\n"\n'
                 '    "<element name=\\"x\\">\\n"\n'
-                '    "    <data type=\\"int\\" />\\n"\n'
+                '    "    <data type=\\"positiveInteger\\" />\\n"\n'
                 '    "</element>\\n"\n'
                 "FREAK OUT\n"
+                '    "</interleave>\\n"\n    ;\n')
+    yield assert_equal, exp_impl, impl
+
+    # schema type tests
+    yield assert_equal, 'string', f._type('std::string')
+    yield assert_equal, 'boolean', f._type('bool')
+    yield assert_equal, 'token', f._type('std::string', 'token')
+
+    m.context = {"MyFactory": OrderedDict([('vars', OrderedDict([ 
+            ('x', {'type': ('std::map', 'int', 'double')}), 
+            ]))
+            ])}
+    impl = f.impl()
+    exp_impl = ('  return ""\n'
+                '    "<interleave>\\n"\n'
+                '    "<element name=\\"x\\">\\n"\n'
+                '    "    <oneOrMore>\\n"\n'
+                '    "        <element name=\\"key\\">\\n"\n'
+                '    "            <data type=\\"int\\" />\\n"\n'
+                '    "        </element>\\n"\n'
+                '    "        <element name=\\"val\\">\\n"\n'
+                '    "            <data type=\\"double\\" />\\n"\n'
+                '    "        </element>\\n"\n'
+                '    "    </oneOrMore>\\n"\n'
+                '    "</element>\\n"\n'
                 '    "</interleave>\\n"\n    ;\n')
     yield assert_equal, exp_impl, impl
 

